build-scripts/build_compiler.js

```diff
@@ -0,0 +1,144 @@
+const packageVersion = new Semver(packageInfo.version);
+
+const fs = require('fs');
+const glob = require('glob');
+
+const gitTagResults = spawnSync('git', ['tag', '--points-at', 'HEAD'], {
+  cwd: './compiler'
+});
+let compilerVersion = '1.0-SNAPSHOT';
+if (gitTagResults.status === 0) {
+  const currentTag = gitTagResults.stdout.toString().replace(/\s/g, '');
+  let normalizedTag = currentTag;
+  if (normalizedTag) {
+    // Standard release tags are of the form "vYYYYMMDD".
+    // We also recognize the form where a special release name proceeds it:
+    //     "webpack-v20180810"
+    //     "v20180810-gwt-fix"
+    normalizedTag = currentTag.replace(/^([a-z]+-)?(v\d{8})(.*)$/, '$2');
+  }
+
+  // If the compiler submodule is pointing at a tagged version that matches the package
+  // major version, update the compiler build to use the correct version number.
+  if (normalizedTag === `v${packageVersion.major}`) {
+    compilerVersion = normalizedTag;
+    const pomPaths = glob.sync('./compiler/**/pom*.xml');
+    pomPaths.forEach(pomPath => {
+      const contents = fs.readFileSync(pomPath, 'utf8');
+      fs.writeFileSync(pomPath, contents.replace('<version>1.0-SNAPSHOT</version>', `<version>${compilerVersion}</version>`), 'utf8');
+    });
+  }
+}
+
+const compilerJavaBinaryPath = `./compiler/target/closure-compiler-${compilerVersion}.jar`;
+const compilerJsBinaryPath = `./compiler/target/closure-compiler-gwt-${compilerVersion}/jscomp/jscomp.js`;
+
+console.log(process.platform, process.arch, compilerVersion);
+
+function copyCompilerBinaries() {
+  ncp(compilerJavaBinaryPath, './packages/google-closure-compiler-java/compiler.jar', err => {
+    if (err) {
+      throw new Error(err);
+    }
+  });
+  ncp(compilerJavaBinaryPath, './packages/google-closure-compiler-linux/compiler.jar', err => {
+    if (err) {
+      throw new Error(err);
+    }
+  });
+  ncp(compilerJavaBinaryPath, './packages/google-closure-compiler-osx/compiler.jar', err => {
+    if (err) {
+      throw new Error(err);
+    }
+  });
+  ncp(compilerJsBinaryPath, './packages/google-closure-compiler-js/jscomp.js', err => {
+    if (err) {
+      throw new Error(err);
+    }
+  });
+  ncp('./compiler/contrib', './packages/google-closure-compiler/contrib', err => {
+    if (err) {
+      throw new Error(err);
+    }
+  });
+}
+
+if (!fs.existsSync(compilerJavaBinaryPath) || !fs.existsSync(compilerJsBinaryPath)) {
+  spawnSync('mvn', ['clean'], {cwd: './compiler', stdio: 'inherit'});
+  const compilerBuild = spawn(
+      'mvn',
+      [
+        '-DskipTests',
+        '-pl',
+        'externs/pom.xml,pom-main.xml,pom-main-shaded.xml,pom-gwt.xml',
+        'install'
+      ],
+      {
+        cwd: './compiler',
+        stdio: 'inherit',
+      });
+  compilerBuild.on('error', err => {
+    throw err;
+  });
+  compilerBuild.on('close', exitCode => {
+    if (exitCode != 0) {
+      process.exit(1);
+      return;
+    }
+    fs.writeFileSync(
```

Comment:
The logic of this file write is confusing since it reads from the same file it's writing.
I presume it works, but it's not obvious that it will.
Could you read the contents of the file separately first, then write it with the comment block prepended?
Thanks.